code/metrics.py

```diff
@@ -0,0 +1,108 @@
+import numpy as np
+
+
+def hits_at_k(y_true, y_pred=None, k=None, is_l2=False):
+    """
+    Given distances between expected email representation and actual email representation,
+    or distribution of the probability that a sender receiver pair has sent a message
+    Hits @ k tells us that of the most confident k predictions, how many of them were correct
+    :param y_true: ndarray having shape (num_examples, ) indicating whether the mail was sent by the s-r pair
+    :param y_pred: ndarray having shape (num_examples, ) containing scores or probabilities
+    :param k: custom value for k; if this is not set, it considers the entire array
+    :param is_l2: flag to indicate whether the y_pred values are scores/errors (True) or probabilities (False)
+    :return: <int> the hits @ k metric
+    """
+    if k is None:
+        k = len(y_true)
+    else:
+        k = min(k, len(y_true))
+
+    if is_l2:
+        sort_order = y_pred.argsort()
+        y_true = y_true[sort_order]
+        correct = 0
+        for i in xrange(k):
+            correct += y_true[i]
+        return correct, k
+    else:
+        # This does not have clear semantics if we are given the binary probabilities
+        return np.sum(y_true[y_pred > 0.5]), k
+
+
+def average_precision_at_k(y_true, y_pred=None, k=None, is_l2=False):
+    """
+    Given distances between expected email representation and actual email representation,
+    or distribution of the probability that a sender receiver pair has sent a message
+    Average precision at k tells us that of the most confident k predictions, how many of them were correct
+    :param y_true: ndarray having shape (num_examples, ) indicating whether the mail was sent by the s-r pair
+    :param y_pred: ndarray having shape (num_examples, ) containing scores or probabilities
+    :param k: custom value for k; if this is not set, it considers the entire array
+    :param is_l2: flag to indicate whether the y_pred values are scores/errors (True) or probabilities (False)
+    :return: <float> the average precision @ k
+    """
+    if k is None:
+        k = len(y_true)
+    else:
+        k = min(k, len(y_true))
+
+    if is_l2:
+        sort_order = y_pred.argsort()
+    else:
+        sort_order = (-y_pred).argsort()
+
+    y_true = y_true[sort_order]
+    correct = 0
+    ap = 0.0
+    for i in xrange(k):
+        correct += y_true[i]
+        precision = correct / (i+1.0)
+        ap += precision * y_true[i]
+    return ap / np.sum(y_true), k
```

Comment:
shouldn't this be ap/k
Think about this, if we get everything right and k is 3 and y_true has 10 elements out of which 6 are 1. You AP@k will be 3/6 = 0.5, while it should have been 1
Reference: https://github.com/benhamner/Metrics/blob/master/Python/ml_metrics/average_precision.py

Reply:
Different people use different notations. This is partly why it took a long time to write this because different views contradicted each other.

This is what I followed:
https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Average_precision

For the example you are describing, we cannot say much unless we know the relative ordering of the elements
Say it is
a) 1000011111, k=3
Looking at the first three docs, we get (1/1 * 1 + 1/2 * 0 + 1/3 * 0) and then should do a by 3
b) 1100001111, k=3
Looking at the first three docs, we get (1/1 * 1 + 2/2 * 1 + 2/3 * 0) and then should do a by 3

Now the /3 does not come because of the just the value of k
There were other documents that should have been selected preferentially.

That being said, I think there still is a small mistake
It should be min(k, np.sum(y_true))

Reply:
which is k by ur definition in line 43 to 46

Reply:
No there is a subtle difference.
min(k, len(y_true)) is to handle asking for precision@20 when we have only 10 documents

min(k, np.sum(y_true)) is so that we don't penalize a person if they give us all the relevant documents
Eg. there were 4 relevant documents, we asked for 10